src/engraving/libmscore/instrument.cpp

```diff
@@ -1739,31 +1739,48 @@ void Instrument::updateInstrumentId()
     // because there are multiple instrument using this same id.
     // For these instruments, use the value of controller 32 of the "arco"
     // channel to find the correct instrument.
+    // There are some duplicate MusicXML IDs among other instruments too. In
+    // that case we check the pitch range and use the shortest ID that matches.
     const QString arco = QString("arco");
     const bool groupHack = instrumentId() == QString("strings.group");
     const int idxref = channelIdx(arco);
     const int val32ref = (idxref < 0) ? -1 : channel(idxref)->bank();
     QString fallback;
+    int bestMatchStrength = 0; // higher when fallback ID provides better match for instrument data
 
     for (InstrumentGroup* g : qAsConst(instrumentGroups)) {
         for (InstrumentTemplate* it : qAsConst(g->instrumentTemplates)) {
-            if (it->musicXMLid == instrumentId()) {
-                if (groupHack) {
-                    if (fallback.isEmpty()) {
-                        // Instrument "Strings" doesn't have bank defined so
-                        // if no "strings.group" instrument with requested bank
-                        // is found, assume "Strings".
-                        fallback = it->id;
-                    }
-                    for (const Channel& chan : it->channel) {
-                        if ((chan.name() == arco) && (chan.bank() == val32ref)) {
-                            _id = it->id;
-                            return;
-                        }
+            if (it->musicXMLid != instrumentId()) {
+                continue;
+            }
+            if (groupHack) {
+                if (fallback.isEmpty()) {
+                    // Instrument "Strings" doesn't have bank defined so
+                    // if no "strings.group" instrument with requested bank
+                    // is found, assume "Strings".
+                    fallback = it->id;
+                }
+                foreach (const Channel& chan, it->channel) {
+                    if ((chan.name() == arco) && (chan.bank() == val32ref)) {
+                        _id = it->id;
+                        return;
                     }
-                } else {
-                    _id = it->id;
-                    return;
+                }
+            } else {
+                int matchStrength = 0
+                                    + (minPitchP() == it->minPitchP)
+                                    + (minPitchA() == it->minPitchA)
+                                    + (maxPitchA() == it->maxPitchA)
+                                    + (maxPitchP() == it->maxPitchP);
```

Comment:
Using pitches is a little tricky. E.g. the lowest pitch for a baritone saxophone is a Bb (written) for amateur and A (written) for professional. This isn't 100% true, the lowest note in this case is defined by the instrument, not the capability of the player (for the low A the instrument needs an extension). Same for Bb clarinets which can have an extension to Eb (instead of having an E as lowest note).
When users decide to change the range this test will fail (BTW, when I have to write for instruments I'm not that familiar with, I actually change the ranges, amateur to what is comfortable to play, professional to what is is possible for the particular player. Now MuseScore gives an indication whether the part is playable for the player I'm writing for).
So, if pitches are used, the `==` isn't that great choice. But whether `<=` or `>=` should be used, no idea.
Maybe the MIDI channel can be used too?

Reply:
We can check against the MIDI program and MIDI bank number for the default channel (i.e. the first channel in the channels list).

> When users decide to change the range this test will fail

The ranges are being used as a weighting to choose between multiple instruments with the same MusicXML ID. The ranges are not used as a strict condition, so if no instrument matches the given ranges then the code simply returns the first instrument that matches the MusicXML ID (actually, it returns the instrument with the shortest ID that matches the MusicXML ID).

But if we wanted to be really fancy we could do this:

```c++
int matchStrength = 100
    - qAbs(minPitchP() - it->minPitchP)
    - qAbs(minPitchA() - it->minPitchA)
    - qAbs(maxPitchA() - it->maxPitchA)
    - qAbs(maxPitchP() - it->maxPitchP);
```

But this may not work so well for unpitched instruments or instruments with missing pitch data, and I think it's probably not worth the effort to work around those cases.

Reply:
Not checking against MIDI program and bank number for now since the current code works without it. I looked at adding it but it's not elegant unless we get rid of the special handling for strings, so I'm leaving it as it is for now.